### PR TITLE
Fix gitolite ext svc connection

### PR DESCRIPTION
**Pull request by mollylogue**
Add test to demonstrate failure scenario when using `httpcli.ActorTransportOpt` with `httpcli.ExternalTransportOpt`. Remove the use of `httpcli.ActorTransportOpt` for now. 

The root of the problem is that we can't wrap the `http.Transport` multiple times, which is what we're trying to do here. 

Looking into a way to add the actor transport back in without breaking things (may need @bobheadxi help), but this is the quickest fix to get the connection working for now. 

It's worth noting that adding this `httpcli.ExternalTransportOpt` to other unit tests causes a similar breakage. It seems that if it goes before other `httpcli.Opt` functions, then it works fine, but if it comes after, then we see the error that looks like `httpcli.ExternalTransportOpt: http.Client.Transport is not an *http.Transport: *recorder.Recorder` which comes from [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httpcli/client.go?L613) when we try to modify the transport, but this assumes it's a `*http.Transport` and it isn't.

## Test plan
added unit tests. also verifying gitolite connection in local sg deployment.

### CODE REVIEW

File: internal/repos/gitolite_test.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+func TestGitoliteSource(t *testing.T) {
+	testCases := []struct {
+		name string
+	}{
+		{
+			name: "basic",
```
Comment (mollylogue) [inline internal/repos/gitolite_test.go:17]: I figured we might want more test cases, which is why I added this. But if we think this single case is sufficient for now, I'm fine removing.

**Comment by bobheadxi** [comment]
> It seems that if it goes before other httpcli.Opt functions, then it works fine, but if it comes after, then we see the error that looks like httpcli.ExternalTransportOpt: http.Client.Transport is not an *http.Transport: *recorder.Recorder which comes from [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httpcli/client.go?L613) when we try to modify the transport, but this assumes it's a *http.Transport and it isn't.

It looks like we can just replace that type assertion with an assertion for `http.RoundTripper` which would resolve this issue right? I'm hesitant to approve this removal that might have security implications for the sake of what appears to be awkward API design internally

**Comment by mollylogue** [comment]
> > It seems that if it goes before other httpcli.Opt functions, then it works fine, but if it comes after, then we see the error that looks like httpcli.ExternalTransportOpt: http.Client.Transport is not an *http.Transport: *recorder.Recorder which comes from [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httpcli/client.go?L613) when we try to modify the transport, but this assumes it's a *http.Transport and it isn't.
> 
> It looks like we can just replace that type assertion with an assertion for `http.RoundTripper` which would resolve this issue right? I'm hesitant to approve this removal that might have security implications for the sake of what appears to be awkward API design internally
> 
> Could you also expand on what exact issue you run into with:
> 
> > The root of the problem is that we can't wrap the http.Transport multiple times, which is what we're trying to do here.

So, we can't just use an `http.RoundTripper` because the `httpcli.ExternalTransportOpt` modifies the underlying `http.Transport` object and some of its fields.

**Comment by bobheadxi** [comment]
> So, we can't just use an http.RoundTripper because the httpcli.ExternalTransportOpt modifies the underlying http.Transport object and some of its fields.

In the link you provide, https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httpcli/client.go?L613, we don't do anything to the transport except clone it - can we just not error out there? If cloning is required, can we just implement per-transport-type cloning?

**Comment by mollylogue** [comment]
> > So, we can't just use an http.RoundTripper because the httpcli.ExternalTransportOpt modifies the underlying http.Transport object and some of its fields.
> 
> In the link you provide, https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httpcli/client.go?L613, we don't do anything to the transport except clone it - can we just not error out there? If cloning is required, can we just implement per-transport-type cloning?

If you take a look at the underlying transport functionality [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httpcli/external.go?L55) it modifies fields that are only present on an `http.Transport` object. So it needs that base transport rather than a Roundtripper

**Comment by mollylogue** [comment]
> > > So, we can't just use an http.RoundTripper because the httpcli.ExternalTransportOpt modifies the underlying http.Transport object and some of its fields.
> > 
> > 
> > In the link you provide, https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httpcli/client.go?L613, we don't do anything to the transport except clone it - can we just not error out there? If cloning is required, can we just implement per-transport-type cloning?
> 
> If you take a look at the underlying transport functionality [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httpcli/external.go?L55) it modifies fields that are only present on an `http.Transport` object. So it needs that base transport rather than a Roundtripper

If you look at the comment [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httpcli/client.go?L113-116) it explains it better than I have. The problem is when there are other `Opt` functions added (as in this case w/ Gitolite), the new functions that are added are called first (see [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/httpcli/client.go?L193-195))

**Comment by mollylogue** [comment]
Closing in favor of https://github.com/sourcegraph/sourcegraph/pull/39775

**Pull request closed without merging**